lib/seahorn/BvOpSem2.cc

```diff
@@ -664,20 +663,47 @@ class OpSemMemManager {
   Expr loadIntFromMem(PtrTy ptr, Expr memReg, unsigned byteSz, uint64_t align) {
     Expr mem = m_ctx.read(memReg);
     SmallVector<Expr, 16> words;
-    // -- read all words
-    for (unsigned i = 0; i < byteSz; i += wordSzInBytes()) {
-      words.push_back(loadAlignedWordFromMem(ptrAdd(ptr, i), mem));
+    int offsetBits = 0;
+    switch (wordSzInBytes()) {
+      // cases where ptrs are known to use a certain number of bits to denote byte offset
+      //   and the rest to denote word aligned address
+      case 2: offsetBits = 1; break;
+      case 4: offsetBits = 2; break;
+      case 8: offsetBits = 3; break;
+    }
+    if (offsetBits != 0 && align % wordSzInBytes() != 0) {
+      for (unsigned i = 0; i < byteSz; i++) {
+        // pointers are partitioned into word address (high bits) and offset (low bits)
+        PtrTy wordAddress = bv::extract(ptrSzInBits() - 1, offsetBits, ptrAdd(ptr, i)),
+              byteOffset = bv::extract(offsetBits - 1, 0, ptrAdd(ptr, i));
+
+        // aligned ptr is current (ptr + i) with offset bits truncated to 0
+        PtrTy alignedPtr = bv::concat(wordAddress, bv::bvnum(0L, offsetBits, ptr->efac()));
+        Expr alignedWord = loadAlignedWordFromMem(alignedPtr, mem);
+
+        byteOffset = bv::zext(byteOffset, wordSzInBits() - 3);
+        // (x << 3) to get bit offset; zero extend to maintain word size
+        PtrTy bitOffset = bv::concat(byteOffset, bv::bvnum(0, 3, ptr->efac()));
+
+        Expr byteOfWord = bv::extract(7, 0, mk<BLSHR>(alignedWord, bitOffset));
```

Comment:
Can this also be encoded in some other way without BLSHR? This should be factored out so that other encodings can be tried as well.